src/lib/Cashlink.ts

```diff
@@ -0,0 +1,453 @@
+                message = '';
+            } else {
+                const messageLength = buf.readUint8();
+                const messageBytes = buf.read(messageLength);
+                message = Utf8Tools.utf8ByteArrayToString(messageBytes);
+            }
+
+            return new Cashlink(
+                keyPair,
+                keyPair.publicKey.toAddress(),
+                value,
+                message,
+                CashlinkState.UNKNOWN);
+        } catch (e) {
+            console.error('Error parsing Cashlink:', e);
+            return null;
+        }
+    }
+
+    public static fromObject(object: CashlinkEntry): Cashlink {
+        return new Cashlink(
+            Nimiq.KeyPair.unserialize(new Nimiq.SerialBuffer(object.keyPair)),
+            Nimiq.Address.fromUserFriendlyAddress(object.address),
+            object.value,
+            object.message,
+            object.state,
+            object.date,
+            object.contactName,
+        );
+    }
+
+    private _network: Promise<NetworkClient>;
+    private _networkClientResolver!: (client: NetworkClient) => void;
+    private _balanceRequest: Promise<number> | null = null;
+    private _currentBalance: number = 0;
+    private _immutable: boolean;
+    private _eventListeners: {[type: string]: Array<(data: any) => void>} = {};
+    private _messageBytes: Uint8Array = new Uint8Array(0);
+    private _value: number | null = null;
+    private _knownTransactions: DetailedPlainTransaction[] = [];
+
+    constructor(
+        public keyPair: Nimiq.KeyPair,
+        public address: Nimiq.Address,
+        value?: number,
+        message?: string,
+        public state: CashlinkState = CashlinkState.UNCHARGED,
+        public date: number = Math.floor(Date.now() / 1000),
+        public contactName?: string, /** unused for now */
+    ) {
+        this._network = new Promise((resolve) => {
+            this._networkClientResolver = resolve;
+        });
+
+        if (value) this.value = value;
+        if (message) this.message = message;
+
+        this._immutable = !!(value || message);
+
+        this._network.then((network: NetworkClient) => {
+            // value will be updated as soon as we have consensus (in _onPotentialBalanceChange)
+            // and a confirmed-amount-changed event gets fired
+            if (network.consensusState === 'established') {
+                this.getAmount(); // Updates _currentBalance internally
+            }
+
+            network.on(NetworkClient.Events.TRANSACTION_PENDING, this._onTransactionAddedOrRelayed.bind(this));
+            network.on(NetworkClient.Events.TRANSACTION_RELAYED, this._onTransactionAddedOrRelayed.bind(this));
+            network.on(NetworkClient.Events.HEAD_CHANGE, this._onHeadChanged.bind(this));
+            network.on(NetworkClient.Events.CONSENSUS_ESTABLISHED, this._onPotentialBalanceChange.bind(this));
+
+            network.subscribe(this.address.toUserFriendlyAddress());
```

Comment:
The lib was changed to also detect re-use by re-charging, so the listener needs to stay I'd say.

Reply:
I think it's ok if we don't assign / keep all these event handler running for claimed cashlinks only because someone might use them in a way that is not intended.
Recharged cashlinks will still be detected on reload.

Reply:
Well, that's _our_ use case.

Reply:
Removed all direct balance-checking from the Cashlink lib and only rely on the BALANCES_CHANGED event from the network iframe. We will stay subscribed to handle recharging events.